File: codes/cpp/chapter_tree/array_binary_tree.cpp
/**
 * File: array_binary_tree.cpp
 * Created Time: 2026-05-19
 * Author: miaochengyou1119(https://github.com/miaochengyou1119)
 * Improved: 现代化 C++ 风格、类型安全、鲁棒性、可读性
 */

#include "../utils/common.hpp"
#include <vector>
#include <climits>
#include <string>
#include <stdexcept>

// 使用枚举替代字符串，类型安全、效率更高
enum class TraversalOrder {
    PRE_ORDER,
    IN_ORDER,
    POST_ORDER
};

/* 数组表示下的二叉树类 */
class ArrayBinaryTree {
public:
    //  explicit 禁止隐式转换，更安全
    explicit ArrayBinaryTree(const std::vector<int>& arr) : tree(arr) {}

    // 成员函数标记为 const，不修改成员变量
    size_t size() const {
        return tree.size();
    }

    // 获取索引为 i 节点的值，越界/空节点返回空值标识
    int val(size_t i) const {
        if (i >= size())
            return INT_MAX;
        return tree[i];
    }

    // 子节点 / 父节点索引计算，统一使用 size_t
    size_t left(size_t i) const {
        return 2 * i + 1;
    }

    size_t right(size_t i) const {
        return 2 * i + 2;
    }

    size_t parent(size_t i) const {
        return (i - 1) / 2;
    }

    // 层序遍历
    std::vector<int> levelOrder() const {
        std::vector<int> res;
        for (size_t i = 0; i < size(); ++i) {
            if (val(i) != INT_MAX)
                res.push_back(val(i));
        }
        return res;
    }

    // 前/中/后序遍历
    std::vector<int> preOrder() const {
        std::vector<int> res;
        dfs(0, TraversalOrder::PRE_ORDER, res);
        return res;
    }

    std::vector<int> inOrder() const {
        std::vector<int> res;
        dfs(0, TraversalOrder::IN_ORDER, res);
        return res;
    }

    std::vector<int> postOrder() const {
        std::vector<int> res;
        dfs(0, TraversalOrder::POST_ORDER, res);
        return res;
    }

private:
    std::vector<int> tree;

    // 私有 DFS，const 安全、枚举类型、无字符串比较
    void dfs(size_t i, TraversalOrder order, std::vector<int>& res) const {
        int node_val = val(i);
        if (node_val == INT_MAX)
            return;

        if (order == TraversalOrder::PRE_ORDER)
            res.push_back(node_val);

        dfs(left(i), order, res);

        if (order == TraversalOrder::IN_ORDER)
            res.push_back(node_val);

        dfs(right(i), order, res);

        if (order == TraversalOrder::POST_ORDER)
            res.push_back(node_val);
    }
};

/* Driver Code */
int main() {
    // 初始化二叉树
    const std::vector<int> arr = {
        1, 2, 3, 4, INT_MAX, 6, 7, 8, 9,
        INT_MAX, INT_MAX, 12, INT_MAX, INT_MAX, 15
    };

    TreeNode* root = vectorToTree(arr);
    std::cout << "\n初始化二叉树\n";
    std::cout << "二叉树的数组表示：\n";
    printVector(arr);
    std::cout << "二叉树的链表表示：\n";
    printTree(root);

    // 数组表示二叉树
    const ArrayBinaryTree abt(arr);

    // 访问节点
    const size_t i = 1;
    const size_t l = abt.left(i);
    const size_t r = abt.right(i);
    const size_t p = abt.parent(i);

    std::cout << "\n当前节点的索引为 " << i
              << "，值为 " << abt.val(i) << "\n";

    auto print_node = [&](size_t idx, const std::string& name) {
        if (abt.val(idx) != INT_MAX)
            std::cout << "其" << name << "子节点的索引为 " << idx
                      << "，值为 " << abt.val(idx) << "\n";
        else
            std::cout << "其" << name << "子节点的索引为 " << idx
                      << "，值为 nullptr\n";
    };

    print_node(l, "左");
    print_node(r, "右");
    std::cout << "其父节点的索引为 " << p
              << "，值为 " << abt.val(p) << "\n";

    // 遍历
    std::vector<int> res = abt.levelOrder();
    std::cout << "\n层序遍历为： ";
    printVector(res);

    res = abt.preOrder();
    std::cout << "前序遍历为： ";
    printVector(res);

    res = abt.inOrder();
    std::cout << "中序遍历为： ";
    printVector(res);

    res = abt.postOrder();
    std::cout << "后序遍历为： ";
    printVector(res);

    return 0;
}

File: ru/codes/rust/chapter_tree/array_binary_tree.rs
/*
 * File: array_binary_tree.rs
 * Created Time: 2023-07-25
 * Author: night-cruise (2586447362@qq.com)
 */

use hello_algo_rust::include::{print_util, tree_node};

/* Класс двоичного дерева в массивном представлении */
struct ArrayBinaryTree {
    tree: Vec<Option<i32>>,
}

impl ArrayBinaryTree {
    /* Конструктор */
    fn new(arr: Vec<Option<i32>>) -> Self {
        Self { tree: arr }
    }

    /* Вместимость списка */
    fn size(&self) -> i32 {
        self.tree.len() as i32
    }

    /* Получить значение узла с индексом i */
    fn val(&self, i: i32) -> Option<i32> {
        // Если индекс выходит за границы, вернуть None, обозначающий пустую позицию
        if i < 0 || i >= self.size() {
            None
        } else {
            self.tree[i as usize]
        }
    }

    /* Получить индекс левого дочернего узла узла с индексом i */
    fn left(&self, i: i32) -> i32 {
        2 * i + 1
    }

    /* Получить индекс правого дочернего узла узла с индексом i */
    fn right(&self, i: i32) -> i32 {
        2 * i + 2
    }

    /* Получить индекс родительского узла узла с индексом i */
    fn parent(&self, i: i32) -> i32 {
        (i - 1) / 2
    }

    /* Обход в ширину */
    fn level_order(&self) -> Vec<i32> {
        self.tree.iter().filter_map(|&x| x).collect()
    }

    /* Обход в глубину */
    fn dfs(&self, i: i32, order: &'static str, res: &mut Vec<i32>) {
        if self.val(i).is_none() {
            return;
        }
        let val = self.val(i).unwrap();
        // Предварительный обход
        if order == "pre" {
            res.push(val);
        }
        self.dfs(self.left(i), order, res);
        // Симметричный обход
        if order == "in" {
            res.push(val);
        }
        self.dfs(self.right(i), order, res);
        // Обратный обход
        if order == "post" {
            res.push(val);
        }
    }

    /* Предварительный обход */
    fn pre_order(&self) -> Vec<i32> {
        let mut res = vec![];
        self.dfs(0, "pre", &mut res);
        res
    }

    /* Симметричный обход */
    fn in_order(&self) -> Vec<i32> {
        let mut res = vec![];
        self.dfs(0, "in", &mut res);
        res
    }

    /* Обратный обход */
    fn post_order(&self) -> Vec<i32> {
        let mut res = vec![];
        self.dfs(0, "post", &mut res);
        res
    }
}

/* Driver Code */
fn main() {
    // Инициализировать двоичное дерево
    // Здесь используется функция, напрямую строящая двоичное дерево из массива
    let arr = vec![
        Some(1),
        Some(2),
        Some(3),
        Some(4),
        None,
        Some(6),
        Some(7),
        Some(8),
        Some(9),
        None,
        None,
        Some(12),
        None,
        None,
        Some(15),
    ];

    let root = tree_node::vec_to_tree(arr.clone()).unwrap();
    println!("\nИнициализация двоичного дерева\n");
    println!("Массивное представление двоичного дерева:");
    println!(
        "[{}]",
        arr.iter()
            .map(|&val| if let Some(val) = val {
                format!("{val}")
            } else {
                "null".to_string()
            })
            .collect::<Vec<String>>()
            .join(", ")
    );
    println!("Связное представление двоичного дерева:");
    print_util::print_tree(&root);

    // Класс двоичного дерева в массивном представлении
    let abt = ArrayBinaryTree::new(arr);

    // Доступ к узлу
    let i = 1;
    let l = abt.left(i);
    let r = abt.right(i);
    let p = abt.parent(i);
    println!(
        "\nТекущий индекс узла = {}, значение = {}",\ni,\nif let Some(val) = abt.val(i) {\n    format!("{val}")\n} else {\n    "null".to_string()\n}
    );
    println!(
        "Индекс левого дочернего узла = {}, значение = {}",
        l,
        if let Some(val) = abt.val(l) {
            format!("{val}")
        } else {
            "null".to_string()
        }
    );
    println!(
        "Индекс правого дочернего узла = {}, значение = {}",
        r,
        if let Some(val) = abt.val(r) {
            format!("{val}")
        } else {
            "null".to_string()
        }
    );
    println!(
        "Индекс родительского узла = {}, значение = {}",
        p,
        if let Some(val) = abt.val(p) {
            format!("{val}")
        } else {
            "null".to_string()
        }
    );

    // Обходить дерево
    let mut res = abt.level_order();
    println!("\nОбход в ширину: {:?}", res);
    res = abt.pre_order();
    println!("Предварительный обход: {:?}", res);
    res = abt.in_order();
    println!("Симметричный обход: {:?}", res);
    res = abt.post_order();
    println!("Обратный обход: {:?}", res);
}
